File: manashelper-bot/src/main/java/kg/manasuniversity/usbtypec/manashelper/service/CourseService.java
package kg.manasuniversity.usbtypec.manashelper.service;

import kg.manasuniversity.usbtypec.manashelper.entity.Course;
import kg.manasuniversity.usbtypec.manashelper.entity.User;
import kg.manasuniversity.usbtypec.manashelper.exception.CourseNotFoundException;
import kg.manasuniversity.usbtypec.manashelper.exception.UserNotFoundException;
import kg.manasuniversity.usbtypec.manashelper.model.CourseSummary;
import kg.manasuniversity.usbtypec.manashelper.repository.CourseRepository;
import kg.manasuniversity.usbtypec.manashelper.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CourseService {
    private final CourseRepository courseRepository;
    private final UserRepository userRepository;

    public List<CourseSummary> getCoursesByDepartment(UUID departmentId, Long userId) {
        List<Course> courses = courseRepository.findAllByDepartmentId(departmentId);
        Set<Integer> trackedCourseIds = getTrackedCourseIds(userId);
        return toSummaries(courses, trackedCourseIds);
    }

    @Transactional
    public List<CourseSummary> toggleTrackedCourse(int courseId, Long userId) {
        Course course = courseRepository.findById(courseId)
            .orElseThrow(() -> new CourseNotFoundException("Course not found with id: " + courseId));
        User user = userRepository.findByIdWithCourses(userId)
            .orElseThrow(() -> new UserNotFoundException("User not found with id: " + userId));

        if (user.getCourses().contains(course)) {
            user.removeCourse(course);
        } else {
            user.addCourse(course);
        }
        userRepository.save(user);

        List<Course> siblingCourses = courseRepository.findAllByDepartmentId(course.getDepartment().getId());
        Set<Integer> trackedCourseIds = user.getCourses().stream().map(Course::getId).collect(Collectors.toSet());
        return toSummaries(siblingCourses, trackedCourseIds);
    }

    private Set<Integer> getTrackedCourseIds(Long userId) {
        return userRepository.findByIdWithCourses(userId)
            .map(user -> user.getCourses().stream().map(Course::getId).collect(Collectors.toSet()))
            .orElseGet(Set::of);
    }

    private List<CourseSummary> toSummaries(List<Course> courses, Set<Integer> trackedCourseIds) {
        return courses.stream()
            .map(course -> new CourseSummary(
                course.getId(),
                course.getNumber(),
                trackedCourseIds.contains(course.getId())
            ))
            .toList();
    }
}

File: src/manashelper/services/course_service.py
import uuid
from collections.abc import Sequence
from dataclasses import dataclass

from manashelper.db.models import Course
from manashelper.repositories.course_repository import CourseRepository
from manashelper.repositories.user_repository import UserRepository


class CourseNotFoundError(Exception):
    def __init__(self, course_id: int) -> None:
        super().__init__(f"Course {course_id} not found")
        self.course_id = course_id


class UserNotFoundError(Exception):
    def __init__(self, user_id: int) -> None:
        super().__init__(f"User {user_id} not found")
        self.user_id = user_id


@dataclass(frozen=True, slots=True)
class CourseSummary:
    id: int
    number: int
    is_tracked: bool


class CourseService:
    def __init__(self, course_repository: CourseRepository, user_repository: UserRepository) -> None:
        self._course_repository = course_repository
        self._user_repository = user_repository

    async def get_courses_by_department(self, department_id: uuid.UUID, user_id: int) -> list[CourseSummary]:
        courses = await self._course_repository.get_all_by_department_id(department_id)
        tracked_course_ids = await self._get_tracked_course_ids(user_id)
        return self._to_summaries(courses, tracked_course_ids)

    async def toggle_tracked_course(self, course_id: int, user_id: int) -> list[CourseSummary]:
        course = await self._course_repository.get_by_id(course_id)
        if course is None:
            raise CourseNotFoundError(course_id)

        user = await self._user_repository.get_with_tracked_courses(user_id)
        if user is None:
            raise UserNotFoundError(user_id)

        tracked_course_ids = {tracked_course.id for tracked_course in user.tracked_courses}
        if course_id in tracked_course_ids:
            user.tracked_courses = [
                tracked_course for tracked_course in user.tracked_courses if tracked_course.id != course_id
            ]
        else:
            user.tracked_courses.append(course)

        sibling_courses = await self._course_repository.get_all_by_department_id(course.department_id)
        updated_tracked_ids = {tracked_course.id for tracked_course in user.tracked_courses}
        return self._to_summaries(sibling_courses, updated_tracked_ids)

    async def _get_tracked_course_ids(self, user_id: int) -> set[int]:
        user = await self._user_repository.get_with_tracked_courses(user_id)
        if user is None:
            return set()
        return {tracked_course.id for tracked_course in user.tracked_courses}

    @staticmethod
    def _to_summaries(courses: Sequence[Course], tracked_course_ids: set[int]) -> list[CourseSummary]:
        return [
            CourseSummary(id=course.id, number=course.number, is_tracked=course.id in tracked_course_ids)
            for course in courses
        ]
